### PR TITLE
Fix #2226 - Add pipeline & task status

### LINKED ISSUE #2226: Add workflow status
Based on the discussion on https://github.com/open-metadata/OpenMetadata/pull/2225

We should update the Pipeline/Task entities (to be groomed) to capture the status. The status can be obtained from Airflow `context` as `context['task_instance'].state` and `context['dag_run']._state`.

See example https://github.com/atlanhq/atlan-lineage-airflow/blob/master/atlan/lineage/models/backend.py#L155

Workflow status can be captured in time series fashion like we are doing it for tableProfile https://github.com/open-metadata/OpenMetadata/blob/main/catalog-rest-service/src/main/resources/json/schema/entity/data/table.json#L320 This will be helpful let's say if we capture the last 10-20 status, users can see when and how the workflow is failed This will also be useful for data quality to check the freshness of the data

### PR DESCRIPTION
**Pull request by pmbrull**
### Describe your changes :

This PR prepares the backend of https://github.com/open-metadata/OpenMetadata/issues/2226.

We're modifying the JSON schemas and adding the PUT methods to handle pipeline & task status in a time-series fashion.

We will then work on https://github.com/open-metadata/OpenMetadata/issues/2316 to update the lineage backend and pass this information.

Thanks!

#
### Type of change :

- [ ] Bug fix
- [ ] Improvement
- [x] New feature
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation

#
### Frontend Preview (Screenshots) :
<p align="center">For frontend related change, please link screenshots of your changes preview! Optional for backend related changes.
</p>

#
### Checklist:

- [x] I have read the [**CONTRIBUTING**](https://docs.open-metadata.org/open-source-community/developer) document.
- [ ] I have performed a self-review of my own. 
- [ ] I have tagged my reviewers below.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [x] All new and existing tests passed.

#

### CODE REVIEW

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/PipelineRepository.java [hunk 1]
```diff
@@ -82,10 +90,70 @@ public Pipeline setFields(Pipeline pipeline, Fields fields) throws IOException,
...
+        .entityExtensionDAO()
+        .insert(
+            pipelineId.toString(), "pipeline.pipelineStatus", "pipelineStatus", JsonUtils.pojoToJson(updatedStatus));
+    setFields(pipeline, Fields.EMPTY_FIELDS);
+    return pipeline.withPipelineStatus(getPipelineStatus(pipeline));
+  }
+
+  // Validate if a given task exists in the pipeline
+  private void validateTask(Pipeline pipeline, String taskName) {
+    boolean validTask = false;
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/PipelineRepository.java:145]: There is a Java lambda function called anyMatch which I believe does most of the logic that we have in this function

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/pipelines/PipelineResource.java [hunk 2]
```diff
@@ -365,6 +366,20 @@ public Response createOrUpdate(
     return response.toResponse();
   }
 
+  @PUT
+  @Path("/{id}/status")
+  @Operation(summary = "Add status data", tags = "pipelines", description = "Add status data to the pipeline.")
```
Comment (mithmatt) [inline catalog-rest-service/src/main/java/org/openmetadata/catalog/resources/pipelines/PipelineResource.java:371]: Are we missing @ApiResponse doc?

File: catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json [hunk 3]
```diff
@@ -56,6 +82,29 @@
...
       },
       "required": ["id", "name"]
+    },
+    "pipelineStatus": {
+      "description": "Series of pipeline executions, its status and task status.",
+      "type": "object",
+      "properties": {
+        "executionDate": {
+          "description": "Date where the job was executed.",
+          "$ref": "../../type/basic.json#/definitions/date"
```
Comment (mithmatt) [inline catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json:92]: Should this be UTC timestamp?

We had some challenges in the past with dateTime conversions. UTC timestamp is a long that can be converted into any desirable format / local timezones without losing precision.

File: catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json [hunk 4]
```diff
@@ -5,6 +5,32 @@
   "description": "This schema defines the Pipeline entity. A pipeline enables the flow of data from source to destination through a series of processing steps. ETL is a type of pipeline where the series of steps Extract, Transform and Load the data.",
   "type": "object",
   "definitions": {
+    "statusType": {
+      "javaType": "org.openmetadata.catalog.type.StatusType",
+      "description": "Enum defining the possible Status.",
+      "type": "string",
+      "enum": [
+        "successful",
```
Comment (mithmatt) [inline catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json:13]: Should this be CamelCase? Successful, Failed, Pending ?

Also do we need "javaEnum" for this ?

File: catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json (hunk 3 above)
Comment (pmbrull) [reply catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json:92]: This is a good point, we would be missing more recurrent executions as well. Thanks for pointing it out!

**Comment by pmbrull** [comment]
Thanks, @mithmatt for the pointers, appreciate it 🙏  Applied the changes, let me know if there's anything else you think can be improved!

File: catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json [hunk 5]
```diff
@@ -56,6 +93,29 @@
         }
       },
       "required": ["id", "name"]
+    },
+    "pipelineStatus": {
+      "description": "Series of pipeline executions, its status and task status.",
+      "type": "object",
+      "properties": {
+        "executionDate": {
```
Comment (mithmatt) [inline catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json:101]: If a pipeline executes every 6 hours, then there will be 4 executions on the same day. 

In this case, should we change this to `executionStartTime`

File: catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/PipelineRepository.java [hunk 6]
```diff
@@ -82,10 +88,63 @@ public Pipeline setFields(Pipeline pipeline, Fields fields) throws IOException,
...
+    // validate all the Tasks
+    for (Status taskStatus : pipelineStatus.getTaskStatus()) {
+      validateTask(pipeline, taskStatus.getName());
+    }
+
+    // Add new status
+    storedMapStatus.put(pipelineStatus.getExecutionDate(), pipelineStatus);
+
+    // Prepare to update status
+    List<PipelineStatus> updatedStatus = new ArrayList<>(storedMapStatus.values());
```
[the remark this answers is not in the document]
Comment (pmbrull) [reply catalog-rest-service/src/main/java/org/openmetadata/catalog/jdbi3/PipelineRepository.java:130]: I tried it out and it complains on a type mismatch of `List vs Collection`

File: catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json (hunk 5 above)
Comment (pmbrull) [reply catalog-rest-service/src/main/resources/json/schema/entity/data/pipeline.json:101]: I was thinking about execution date similar to the naming in Airflow, but we could indeed change this

**Review by mithmatt — APPROVED** [review]

**Comment by harshach** [comment]
@pmbrull Thanks for the PR. Couple of things we should add to it in another PR
1. Log link from the tasks. such that users can click and go there to get that error logs directly. When a task or pipeline failed
2. Managing the pipelines from the OM itself, we can provide wrappers around Airflow APIs to trigger or clear run and allow users to re-run. This might be interesting to implement in a way that will allow us to support multiple workflow engines such as airflow, prefect

**Comment by harshach** [comment]
@pmbrull looking at the PR, it looks like you are only adding status at the Pipeline level but we should record at task level as well.

**Review by harshach — APPROVED** [review]

**Pull request merged**